primary/src/grpc_server/validator.rs

```diff
@@ -14,35 +19,89 @@ use tonic::{Request, Response, Status};
 use types::{
     BatchMessageProto, BlockError, BlockRemoverErrorKind, CertificateDigest,
     CertificateDigestProto, CollectionRetrievalResult, Empty, GetCollectionsRequest,
-    GetCollectionsResponse, RemoveCollectionsRequest, Validator,
+    GetCollectionsResponse, ReadCausalRequest, ReadCausalResponse, RemoveCollectionsRequest,
+    Validator,
 };
 
-#[derive(Debug)]
-pub struct NarwhalValidator {
+pub struct NarwhalValidator<
+    PublicKey: VerifyingKey,
+    SynchronizerHandler: Handler<PublicKey> + Send + Sync + 'static,
+> {
     tx_get_block_commands: Sender<BlockCommand>,
     tx_block_removal_commands: Sender<BlockRemoverCommand>,
     get_collections_timeout: Duration,
     remove_collections_timeout: Duration,
+    block_synchronizer_handler: Arc<SynchronizerHandler>,
+    dag: Option<Arc<Dag<PublicKey>>>,
 }
 
-impl NarwhalValidator {
+impl<PublicKey: VerifyingKey, SynchronizerHandler: Handler<PublicKey> + Send + Sync + 'static>
+    NarwhalValidator<PublicKey, SynchronizerHandler>
+{
     pub fn new(
         tx_get_block_commands: Sender<BlockCommand>,
         tx_block_removal_commands: Sender<BlockRemoverCommand>,
         get_collections_timeout: Duration,
         remove_collections_timeout: Duration,
+        block_synchronizer_handler: Arc<SynchronizerHandler>,
+        dag: Option<Arc<Dag<PublicKey>>>,
     ) -> Self {
         Self {
             tx_get_block_commands,
             tx_block_removal_commands,
             get_collections_timeout,
             remove_collections_timeout,
+            block_synchronizer_handler,
+            dag,
         }
     }
 }
 
 #[tonic::async_trait]
-impl Validator for NarwhalValidator {
+impl<PublicKey: VerifyingKey, SynchronizerHandler: Handler<PublicKey> + Send + Sync + 'static>
+    Validator for NarwhalValidator<PublicKey, SynchronizerHandler>
+{
+    async fn read_causal(
+        &self,
+        request: Request<ReadCausalRequest>,
+    ) -> Result<Response<ReadCausalResponse>, Status> {
+        let collection_id = request
+            .into_inner()
+            .collection_id
+            .ok_or_else(|| Status::invalid_argument("Missing collection"))?;
```

Comment:
nit: I would probably make the error message a bit more descriptive, something like:
```
No collection id has been provided
```